File: 2025-03-March-GFG-POTD/March 09 - Palindrome SubStrings/Solution.py
def countPS(s):
    n = len(s)
    dp = [[False] * n for _ in range(n)]
    count = 0

    # Checking substrings of length 1 (ignored in count)
    for i in range(n):
        dp[i][i] = True

    # Checking substrings of length 2
    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            count += 1

    # Checking substrings of length ≥3
    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            if s[i] == s[j] and dp[i + 1][j - 1]:
                dp[i][j] = True
                count += 1

    return count


# Driver Code
t = int(input())
for _ in range(t):
    s = input().strip()
    print(countPS(s))
    print("~")

File: 2025-03-March-GFG-POTD/March 09 - Palindrome SubStrings/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countPS(string s) {
        int n = s.length();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int count = 0;

        // Checking substrings of length 1 (ignored in count)
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        // Checking substrings of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s[i] == s[i + 1]) {
                dp[i][i + 1] = true;
                count++;
            }
        }

        // Checking substrings of length ≥3
        for (int length = 3; length <= n; length++) {
            for (int i = 0; i <= n - length; i++) {
                int j = i + length - 1;
                if (s[i] == s[j] && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    count++;
                }
            }
        }

        return count;
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        Solution obj;
        cout << obj.countPS(s) << endl;
        cout << "~" << endl;
    }
    return 0;
}
